### ISSUE TITLE
MAX_WORD_LENGTH - Issue with length of string in Unicode characters

### ISSUE THREAD

**Issue by muhelen** [cwackerfuss/react-wordle#203]
When using unicode character like "மு" (\u0BAE\u0BC1), the length return as 2, but the "actual"length should be 1.  This affected the MAX_WORD_LENGTH throughout the code.

How to fix this?

**Comment by gheja** [comment]
Unfortunately a "letter" is not always one unicode character and it is hard to parse them properly.

The best to do this would be using a well maintaned library like [grapheme-splitter](https://github.com/orling/grapheme-splitter), copied directly from that readme:
```javascript
var splitter = new GraphemeSplitter();

// plain latin alphabet - nothing spectacular
splitter.splitGraphemes("abcd"); // returns ["a", "b", "c", "d"]

// two-char emojis and six-char combined emoji
splitter.splitGraphemes("🌷🎁💩😜👍🏳️‍🌈"); // returns ["🌷","🎁","💩","😜","👍","🏳️‍🌈"]

// diacritics as combining marks, 10 JavaScript chars
splitter.splitGraphemes("Ĺo͂řȩm̅"); // returns ["Ĺ","o͂","ř","ȩ","m̅"]

// individual Korean characters (Jamo), 4 JavaScript chars
splitter.splitGraphemes("뎌쉐"); // returns ["뎌","쉐"]

// Hindi text with combining marks, 8 JavaScript chars
splitter.splitGraphemes("अनुच्छेद"); // returns ["अ","नु","च्","छे","द"]

// demonic multiple combining marks, 75 JavaScript chars
splitter.splitGraphemes("Z͑ͫ̓ͪ̂ͫ̽͏̴̙̤̞͉͚̯̞̠͍A̴̵̜̰͔ͫ͗͢L̠ͨͧͩ͘G̴̻͈͍͔̹̑͗̎̅͛́Ǫ̵̹̻̝̳͂̌̌͘!͖̬̰̙̗̿̋ͥͥ̂ͣ̐́́͜͞"); // returns ["Z͑ͫ̓ͪ̂ͫ̽͏̴̙̤̞͉͚̯̞̠͍","A̴̵̜̰͔ͫ͗͢","L̠ͨͧͩ͘","G̴̻͈͍͔̹̑͗̎̅͛́","Ǫ̵̹̻̝̳͂̌̌͘","!͖̬̰̙̗̿̋ͥͥ̂ͣ̐́́͜͞"]
```

I have added a PR that aims to fix this issue, along with displaying, entering and deleting the letters: #205

**Comment by gheja** [comment]
A side note/fun fact: there are alphabets where the special letters are made up entirely of regular letters, like in Hungarian there is "sz" which is just "s" + "z" but they are handled as one letter. And not all occurences of "s" followed by "z" are "sz", it depends on the word. Check out [szózat](https://szozat.miklosdanka.com/), the Hungarian fork by @mdanka, supporting these letters.

**Comment by muhelen** [comment]
> I have added a PR that aims to fix this issue, along with displaying, entering and deleting the letters: #205

Hi, thanks for the reply. I've used your solutions https://github.com/cwackerfuss/react-wordle/pull/205, but the letters on grid guess stops after 3 Tamil characters combination. க்ச்ட் count as 5 instead of 3, whereas கசட count as 3 can continue the next 2 characters.

**Comment by mdanka** [comment]
@gheja thanks for the mention :) Yeah, for the Hungarian version we had to separate the concept of characters and of letters, so words became "letter arrays", and letters can contain multiple characters.

![Feb-16-2022 10-53-35](https://user-images.githubusercontent.com/8069999/154250071-b8e7b9bd-0dc8-48f0-8f35-b42bee5f6b61.gif)

**Comment by gheja** [comment]
@muhelen wrote:
> Its still didn't fix the MAX_WORD_LENGTH issue. The unicode characters still count as separate characters, which in return affecting the length number.

@muhelen wrote:
> Now NOT_ENOUGH_LETTERS_MESSAGE working fine. The unicode characters combination could be up to 5 letters, but the last letter couldn't combine fully. For example, it could enter as ம but ம+் which in return supposed to be ம் not working..

How did you enter the letters?

I have modified the keyboard to have the letters you provided and also some copy-pasted from Wikipedia and it looks fine to me, could not reproduce the issue.

![image](https://user-images.githubusercontent.com/1894642/154340792-0036b174-29e2-4a66-8bca-629ce7ffe1ac.png)

(I am working on deploying this version with the modified keyboard to somewhere.)

**Comment by muhelen** [comment]
@gheja It's almost working, only the last character not combining. I've given the correct Tamil keyboard code down here. 

First row =  **dependants** , Second row = **vowels**, Third & forth row = **consonants**

**Vowels** are independent characters, while **consonants** + **dependants** will form **compound** (combine) **consonants**.

You would notice only the last grid character won't compound, but other grids will combine. 

<img width="574" alt="error" src="https://user-images.githubusercontent.com/5513469/154385740-85a953f8-6cbf-4a81-9ce0-08aadccc5d94.png">

```
<div>
      <div className="flex justify-center mb-1">
        {['்','ா', 'ி', 'ீ', 'ு', 'ூ', 'ெ', 'ே', 'ை', 'ொ','ோ','ௌ'].map((key) => (
          <Key
            value={key}
            key={key}
            onClick={onClick}
            status={charStatuses[key]}
            isRevealing={isRevealing}
          />
        ))}
      </div>
      <div className="flex justify-center mb-1">
        {['அ', 'ஆ', 'இ', 'ஈ','உ', 'ஊ', 'எ', 'ஏ','ஐ', 'ஒ', 'ஓ', 'ஔ'].map((key) => (
          <Key
            value={key}
            key={key}
            onClick={onClick}
            status={charStatuses[key]}
            isRevealing={isRevealing}
          />
        ))}
      </div>
      <div className="flex justify-center mb-1">
        {['க', 'ச', 'ட', 'த', 'ப', 'ற','ங', 'ஞ', 'ண', 'ந', 'ம', 'ன'].map((key) => (
          <Key
            value={key}
            key={key}
            onClick={onClick}
            status={charStatuses[key]}
            isRevealing={isRevealing}
          />
        ))}
      </div>
      <div className="flex justify-center">
      <Key width={65.4} value="ENTER" onClick={onClick}>
          {ENTER_TEXT}
        </Key>
      {['ய', 'ர', 'ல', 'வ', 'ழ', 'ள'].map((key) => (
          <Key
            value={key}
            key={key}
            onClick={onClick}
            status={charStatuses[key]}
            isRevealing={isRevealing}
          />  
        ))}
        <Key width={65.4} value="DELETE" onClick={onClick}>
          {DELETE_TEXT}
        </Key>
          </div>
    </div>
```

**Comment by gheja** [comment]
Oh, so you have to press two buttons to get one letter.

I believe then the problem is that the code checks how many characters already present and will refuse to add another.

We have to allow a new character if that's a modifier. It might be a challenge as - if I remember correctly - in Unicode more than one modifier can be applied to a letter but if an invalid combination is tried then they becomes two separate letters.

I think I have an idea - don't check the length of the "current guess" but check the length of the "current guess + new input" and limit that. I'll try to implement this change soon.

**Comment by muhelen** [comment]
@gheja Vowels and Consonants could be input with single button press, but most of the other characters needed two button combinations.
Your idea on the length check make sense. Thanks for your valuable contribution on this issue.

**Comment by gheja** [comment]
I made the change in the code and tested with the keyboard you provided. I think it is working as expected, however the modifiers cannot be deleted separately but the whole letter+modifier. Updated my pull request (#205).

Thank you for the kind words :) and also for the great reporting and the examples, they were really helpful. I hope this fixes the issue but please let me know if still persists.

**Comment by muhelen** [comment]
@gheja Great. It's working fine on all the grids now 😊 
Yes the modifiers cannot be deleted separately, but for Tamil letter it's fine. Not sure on other language characters usage.

Thank you very much for your great contribution to fix on this issue 🙏🏼🙏🏼🙏🏼. All the best for all your endeavours.👍🏼👍🏼

**Comment by muhelen** [comment]
@gheja I've found your another code on previous CompletedRow.tsx file working correctly. 

`import { getGuessStatuses } from '../../lib/statuses'
import { Cell } from './Cell'
import GraphemeSplitter from 'grapheme-splitter'

type Props = {
  guess: string
  isRevealing?: boolean
}

export const CompletedRow = ({ guess, isRevealing }: Props) => {
  const statuses = getGuessStatuses(guess)
  const splitter = new GraphemeSplitter()

  return (
    <div className="flex justify-center mb-1">
      {splitter.splitGraphemes(guess).map((letter, i) => (
        <Cell
          key={i}
          value={letter}
          status={statuses[i]}
          position={i}
          isRevealing={isRevealing}
          isCompleted
        />
      ))}
    </div>
  )
}`

<img width="571" alt="Screenshot 2022-02-19 at முற்பகல் 3 06 13" src="https://user-images.githubusercontent.com/5513469/154746689-0afb1343-8edf-4936-a759-53f97048c473.png">


But, on the keyboard the combined letters reveals correctly not displaying the color correctly. I'll try to check on the characters too..

**Comment by gheja** [comment]
Thanks, I missed that! I have also checked the project and found more string handling/splitting that needed fixing.

> on the keyboard the combined consonants unicode letters (letter+modifier) reveals not displaying the colors correctly. Only the vowels color change.

I have improved it but I believe this can't be done correctly universally, as it is possible to use a modifier for more than one letter, and use a letter with different modifiers (I am not sure if it works in Tamil but for example in Hungarian there are o ó ö ő, u ú ü ű, a á, e é (however in this case there are separate physical buttons for them on the keyboard and they each have distinct unicode characters)).

The code currently marks the colors on the keys if there's a direct match with a letter of the guesses. I.e. if a guess contains a "அ" then the key "அ" will be marked. If the guess contains a "அ்" (which is the same letter but with a modifier so two keys have to be pressed to have this letter) then no key will be marked.

This isn't perfect but better than before and I could not find a solution for this other than disabling markings of keys.

**Comment by muhelen** [comment]
@gheja Great. It's working superb now 😊.  I've checked with the keys modified on the keyboard to show the combine letters and its works perfectly too. 👍🏼

<img width="550" alt="Screenshot 2022-02-19 at பிற்பகல் 9 08 22" src="https://user-images.githubusercontent.com/5513469/154802274-5590aa85-490c-4a5b-b219-7303efff46fa.png">

The only issue now is the grid width not expanding according to the combined characters. The current grid is fixed in width size, it needs to be responsive width according to the size of the compound characters. Some of the mixed characters to test : நௌ, ணௌ, ணை, றை, றோ, றௌ, ஙௌ, ஞோ, ஞே,  ..

**Comment by gheja** [comment]
I'm so glad it works, thanks for the response!

Yes, the grid is strictly a square now and I also noticed the clipping while testing. I think we need to have a configuration option to set it wider so the letters in specific alphabets could fit. I have opened a separate Issue for this: #240

**Comment by muhelen** [comment]
@gheja Thanks again for your codes and help on resolving this issue. 

I'll close this issue now as it's already fix and commit to the main repository. 

Hope the issue https://github.com/cwackerfuss/react-wordle/issues/240 got solutions too. Thanks.
